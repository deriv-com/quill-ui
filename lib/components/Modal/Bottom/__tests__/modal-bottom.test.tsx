import React from "react";
import ReactDOM from "react-dom";
import { render, screen, waitFor } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { StandaloneTrashRegularIcon } from "@deriv/quill-icons";
import { ModalBottom } from "..";

describe("ModalBottom", () => {
    const primaryButtonLabel = "Primary Button Label";
    const secondaryButtonLabel = "Secondary Button Label";
    const shortTextContent = "This is some amazing placeholder.";
    const imageSRC =
        "https://live.staticflickr.com/603/21947667154_e63cc9252b_b.jpg";

    const children = (
        <>
            <ModalBottom.Title>Title</ModalBottom.Title>
            <ModalBottom.Body>{shortTextContent}</ModalBottom.Body>
        </>
    );

    const mockProps = {
        isOpened: true,
        toggleModal: jest.fn(),
        primaryButtonLabel,
    };

    beforeAll(() => {
        (ReactDOM.createPortal as jest.Mock) = jest.fn(
            (component) => component,
        );
    });

    afterAll(() => {
        (ReactDOM.createPortal as jest.Mock).mockClear();
    });

    it("should render with default values if optional ones were not passed", () => {
        const { container } = render(
            <ModalBottom {...mockProps}>{children}</ModalBottom>,
        );

        expect(container).toMatchSnapshot();
    });

    it("should call toggleModal if user clicks on the overlay", async () => {
        render(<ModalBottom {...mockProps}>{children}</ModalBottom>);

        userEvent.click(screen.getByTestId("dt_overlay"));
        await waitFor(() => {
            expect(mockProps.toggleModal).toHaveBeenCalledTimes(1);
        });
    });

    it("should not render handleBar if showHandleBar === false", () => {
        const { container } = render(
            <ModalBottom {...mockProps} showHandleBar={false}>
                {children}
            </ModalBottom>,
        );

        expect(screen.queryByTestId("dt_handlebar")).not.toBeInTheDocument();
        expect(container).toMatchSnapshot();
    });

    it("should render secondary button if showSecondaryButton === true and the label was passed", () => {
        const { container } = render(
            <ModalBottom
                {...mockProps}
                showSecondaryButton
                secondaryButtonLabel={secondaryButtonLabel}
            >
                {children}
            </ModalBottom>,
        );

        expect(screen.getByText(secondaryButtonLabel)).toBeInTheDocument();
        expect(container).toMatchSnapshot();
    });

    it("should call toggleModal if user clicks on secondary button", async () => {
        render(
            <ModalBottom
                {...mockProps}
                showSecondaryButton
                secondaryButtonLabel={secondaryButtonLabel}
            >
                {children}
            </ModalBottom>,
        );

        userEvent.click(screen.getByText(secondaryButtonLabel));
        await waitFor(() => {
            expect(mockProps.toggleModal).toHaveBeenCalledTimes(1);
        });
    });

    it("should call toggleModal if user clicks on primary button and shouldCloseOnPrimaryButtonClick == true", async () => {
        render(
            <ModalBottom {...mockProps} shouldCloseOnPrimaryButtonClick>
                {children}
            </ModalBottom>,
        );

        userEvent.click(screen.getByText(primaryButtonLabel));
        await waitFor(() => {
            expect(mockProps.toggleModal).toHaveBeenCalledTimes(1);
        });
    });

    it("should call primaryButtonCallback if user clicks on primary button and primaryButtonCallback was passed", async () => {
        const primaryButtonCallback = jest.fn();
        render(
            <ModalBottom
                {...mockProps}
                primaryButtonCallback={primaryButtonCallback}
            >
                {children}
            </ModalBottom>,
        );

        userEvent.click(screen.getByText(primaryButtonLabel));
        await waitFor(() => {
            expect(primaryButtonCallback).toHaveBeenCalledTimes(1);
        });
    });

    it("should render passed image if it is ReactNode", () => {
        const { container } = render(
            <ModalBottom {...mockProps}>
                <ModalBottom.Image>
                    <img src={imageSRC} alt="Apples" />
                </ModalBottom.Image>
            </ModalBottom>,
        );

        expect(screen.getByRole("img")).toBeInTheDocument();
        expect(container).toMatchSnapshot();
    });

    it("should render container with image as a background if src was passed", () => {
        const { container } = render(
            <ModalBottom {...mockProps}>
                <ModalBottom.Image src={imageSRC} />
            </ModalBottom>,
        );

        expect(screen.getByTestId("dt_modal_image")).toBeInTheDocument();
        expect(container).toMatchSnapshot();
    });

    it("should render SVG if it was passed", () => {
        const { container } = render(
            <ModalBottom {...mockProps}>
                <ModalBottom.Image>
                    <StandaloneTrashRegularIcon />
                </ModalBottom.Image>
            </ModalBottom>,
        );

        expect(container).toMatchSnapshot();
    });
});
