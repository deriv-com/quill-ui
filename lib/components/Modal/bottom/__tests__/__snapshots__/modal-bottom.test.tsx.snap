// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ModalBottom should not render handleBar if showHandleBar === false 1`] = `
<div>
  <div
    class="quill-modal-bottom__overlay"
    data-testid="dt_overlay"
  >
    <div
      class="quill-modal-bottom__container quill-modal-bottom__container--visible"
    >
      <div
        class="quill-modal-bottom__content-wrapper"
      >
        <h4
          class="quill-typography__h4 quill-typography__color--prominent quill-modal-bottom__content-title"
        >
          Title
        </h4>
        <div
          class="quill-typography__body-text__size--md__weight--regular__decoration--default quill-modal-bottom__content-body"
        >
          This is some amazing placeholder.
        </div>
      </div>
      <div
        class="quill-modal-bottom__button-wrapper"
      >
        <button
          class="quill-button quill-button__size--lg quill__color--primary-black quill-button__full-width"
          data-state=""
        >
          <span
            class="button-label"
          >
            <p
              class="quill-typography__body-text__size--md__weight--bold__decoration--default black"
            >
              Primary Button Label
            </p>
          </span>
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`ModalBottom should render SVG if it was passed 1`] = `
<div>
  <div
    class="quill-modal-bottom__overlay"
    data-testid="dt_overlay"
  >
    <div
      class="quill-modal-bottom__container quill-modal-bottom__container--visible"
    >
      <div
        class="quill-modal-bottom__handle-bar"
        data-testid="dt_handlebar"
      />
      <div
        class="quill-modal-bottom__content-wrapper"
      >
        <div
          class="quill-modal-bottom__content-image"
        >
          <svg
            height="32"
            role="img"
            viewBox="0 0 32 32"
            width="32"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M14.164 7.75a.61.61 0 0 0-.508.313L13.031 9h5.899l-.625-.937a.61.61 0 0 0-.508-.313zM20.375 9h3.75c.313 0 .625.313.625.625a.64.64 0 0 1-.625.625h-.742l-.977 13.945c-.117 1.328-1.172 2.305-2.5 2.305h-7.851a2.51 2.51 0 0 1-2.5-2.305L8.578 10.25h-.703a.617.617 0 0 1-.625-.625c0-.312.273-.625.625-.625h3.711l1.016-1.602a1.84 1.84 0 0 1 1.562-.898h3.633c.625 0 1.25.352 1.601.898zm1.758 1.25H9.828l.977 13.867c.039.625.586 1.133 1.25 1.133h7.851c.664 0 1.211-.508 1.25-1.133z"
            />
          </svg>
        </div>
        <h4
          class="quill-typography__h4 quill-typography__color--prominent quill-modal-bottom__content-title"
        >
          Title
        </h4>
      </div>
      <div
        class="quill-modal-bottom__button-wrapper"
      >
        <button
          class="quill-button quill-button__size--lg quill__color--primary-black quill-button__full-width"
          data-state=""
        >
          <span
            class="button-label"
          >
            <p
              class="quill-typography__body-text__size--md__weight--bold__decoration--default black"
            >
              Primary Button Label
            </p>
          </span>
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`ModalBottom should render container with image as a background if src was passed 1`] = `
<div>
  <div
    class="quill-modal-bottom__overlay"
    data-testid="dt_overlay"
  >
    <div
      class="quill-modal-bottom__container quill-modal-bottom__container--visible"
    >
      <div
        class="quill-modal-bottom__handle-bar"
        data-testid="dt_handlebar"
      />
      <div
        class="quill-modal-bottom__content-wrapper"
      >
        <div
          class="quill-modal-bottom__content-image"
          data-testid="dt_modal_image"
        />
        <h4
          class="quill-typography__h4 quill-typography__color--prominent quill-modal-bottom__content-title"
        >
          Title
        </h4>
      </div>
      <div
        class="quill-modal-bottom__button-wrapper"
      >
        <button
          class="quill-button quill-button__size--lg quill__color--primary-black quill-button__full-width"
          data-state=""
        >
          <span
            class="button-label"
          >
            <p
              class="quill-typography__body-text__size--md__weight--bold__decoration--default black"
            >
              Primary Button Label
            </p>
          </span>
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`ModalBottom should render passed image if it is ReactNode 1`] = `
<div>
  <div
    class="quill-modal-bottom__overlay"
    data-testid="dt_overlay"
  >
    <div
      class="quill-modal-bottom__container quill-modal-bottom__container--visible"
    >
      <div
        class="quill-modal-bottom__handle-bar"
        data-testid="dt_handlebar"
      />
      <div
        class="quill-modal-bottom__content-wrapper"
      >
        <div
          class="quill-modal-bottom__content-image"
        >
          <img
            alt="Apples"
            src="https://live.staticflickr.com/603/21947667154_e63cc9252b_b.jpg"
          />
        </div>
        <h4
          class="quill-typography__h4 quill-typography__color--prominent quill-modal-bottom__content-title"
        >
          Title
        </h4>
      </div>
      <div
        class="quill-modal-bottom__button-wrapper"
      >
        <button
          class="quill-button quill-button__size--lg quill__color--primary-black quill-button__full-width"
          data-state=""
        >
          <span
            class="button-label"
          >
            <p
              class="quill-typography__body-text__size--md__weight--bold__decoration--default black"
            >
              Primary Button Label
            </p>
          </span>
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`ModalBottom should render secondary button if showSecondaryButton === true and the label was passed 1`] = `
<div>
  <div
    class="quill-modal-bottom__overlay"
    data-testid="dt_overlay"
  >
    <div
      class="quill-modal-bottom__container quill-modal-bottom__container--visible"
    >
      <div
        class="quill-modal-bottom__handle-bar"
        data-testid="dt_handlebar"
      />
      <div
        class="quill-modal-bottom__content-wrapper"
      >
        <h4
          class="quill-typography__h4 quill-typography__color--prominent quill-modal-bottom__content-title"
        >
          Title
        </h4>
        <div
          class="quill-typography__body-text__size--md__weight--regular__decoration--default quill-modal-bottom__content-body"
        >
          This is some amazing placeholder.
        </div>
      </div>
      <div
        class="quill-modal-bottom__button-wrapper"
      >
        <button
          class="quill-button quill-button__size--lg quill__color--primary-black quill-button__full-width"
          data-state=""
        >
          <span
            class="button-label"
          >
            <p
              class="quill-typography__body-text__size--md__weight--bold__decoration--default black"
            >
              Primary Button Label
            </p>
          </span>
        </button>
        <button
          class="quill-button quill-button__size--lg quill__color--secondary-black quill-modal-bottom__button quill-button__full-width"
          data-state=""
        >
          <span
            class="button-label"
          >
            <p
              class="quill-typography__body-text__size--md__weight--bold__decoration--default black"
            >
              Secondary Button Label
            </p>
          </span>
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`ModalBottom should render with default values if optional ones were not passed 1`] = `
<div>
  <div
    class="quill-modal-bottom__overlay"
    data-testid="dt_overlay"
  >
    <div
      class="quill-modal-bottom__container quill-modal-bottom__container--visible"
    >
      <div
        class="quill-modal-bottom__handle-bar"
        data-testid="dt_handlebar"
      />
      <div
        class="quill-modal-bottom__content-wrapper"
      >
        <h4
          class="quill-typography__h4 quill-typography__color--prominent quill-modal-bottom__content-title"
        >
          Title
        </h4>
        <div
          class="quill-typography__body-text__size--md__weight--regular__decoration--default quill-modal-bottom__content-body"
        >
          This is some amazing placeholder.
        </div>
      </div>
      <div
        class="quill-modal-bottom__button-wrapper"
      >
        <button
          class="quill-button quill-button__size--lg quill__color--primary-black quill-button__full-width"
          data-state=""
        >
          <span
            class="button-label"
          >
            <p
              class="quill-typography__body-text__size--md__weight--bold__decoration--default black"
            >
              Primary Button Label
            </p>
          </span>
        </button>
      </div>
    </div>
  </div>
</div>
`;
